src/plonk/mod.rs

```diff
@@ -295,27 +304,42 @@ impl<F: PrimeField> PlonkStructure<F> {
             W1s: &W.W,
             W2s: &vec![],
         };
-        let res = (0..nrow)
+
+        (0..total_row)
             .into_par_iter()
-            .map(|row| data.eval(&poly, row))
-            .collect::<Result<Vec<F>, _>>()?
-            .into_iter()
-            .enumerate()
-            .filter(|(i, v)| W.E[*i].ne(v))
-            .count();
+            .map(|row| {
+                data.eval(&poly, row)
+                    .map(|eval_of_row| if eval_of_row.eq(&W.E[row]) { 0 } else { 1 })
+            })
+            .try_reduce(
+                || 0,
+                |mismatch_count, is_missed| Ok(mismatch_count + is_missed),
+            )
+            .map(|mismatch_count| {
+                Some(Error::EvaluationMismatch {
+                    mismatch_count: NonZeroUsize::new(mismatch_count)?,
+                    total_row,
+                })
+            })?
+            .err_or(())?;
+
+        if !self.is_sat_log_derivative(&W.W) {
+            return Err(Error::LogDerivativeNotSat);
+        }
 
-        let is_E_equal = ck.commit(&W.E).unwrap().eq(&U.E_commitment);
-        let is_h_equal_g = self.is_sat_log_derivative(&W.W);
-
-        match (res == 0, check_W_commitments == 0, is_E_equal, is_h_equal_g) {
-            (true, true, true, true) => Ok(()),
-            (false, _, _, _) => Err(Error::EvaluationMismatch {
-                mismatch_count: res,
-                total_row: nrow,
-            }),
-            (_, _, _, false) => Err(Error::LogDerivativeNotSat),
-            _ => Err(Error::CommitmentMismatch),
+        U.W_commitments
```

Comment:
by keep this to the top, will minimize the code diff for review.

Reply:
Answered above